lxd/api_metrics.go

```diff
@@ -331,31 +331,31 @@ func metricsGet(d *Daemon, r *http.Request) response.Response {
 }
 
 func getFilteredMetrics(s *state.State, r *http.Request, compress bool, metricSet *metrics.MetricSet) response.Response {
+	// Ignore filtering in case the authentication for metrics is disabled.
 	if !s.GlobalConfig.MetricsAuthentication() {
 		return response.SyncResponsePlain(true, compress, metricSet.String())
 	}
 
-	// Get instances the user is allowed to view.
-	userHasPermission, err := s.Authorizer.GetPermissionChecker(r.Context(), r, auth.EntitlementCanView, entity.TypeInstance)
-	if err != nil && !auth.IsDeniedError(err) {
+	userHasProjectPermission, err := s.Authorizer.GetPermissionChecker(r.Context(), r, auth.EntitlementCanViewMetrics, entity.TypeProject)
+	if err != nil {
 		return response.SmartError(err)
-	} else if err != nil {
-		// This is counterintuitive. We are unauthorized to get a permission checker for viewing instances because a metric type certificate
-		// can't view instances. However, in order to get to this point we must already have auth.EntitlementCanViewMetrics. So we can view
-		// the metrics but we can't do any filtering, so just return the metrics.
-		return response.SyncResponsePlain(true, compress, metricSet.String())
 	}
 
-	// Filter by project name and instance name.
-	metricSet.FilterSamples(userHasPermission)
-
-	userHasPermission, err = s.Authorizer.GetPermissionChecker(r.Context(), r, auth.EntitlementCanView, entity.TypeProject)
+	userHasServerPermission, err := s.Authorizer.GetPermissionChecker(r.Context(), r, auth.EntitlementCanViewMetrics, entity.TypeServer)
 	if err != nil {
 		return response.SmartError(err)
 	}
 
-	// Filter by project only.
-	metricSet.FilterSamples(userHasPermission)
+	// Filter all metrics for view permissions.
+	// Internal server metrics without labels are compared against the server entity.
+	metricSet.FilterSamples([]func(labels map[string]string) bool{
```

Comment:
Is there a reason you went for `metricSet.FilterSample` taking a slice of `func(labels map[string]string) bool` rather than instead taking just a single `func(labels map[string]string) bool` and then implementing a function that did these two `userHas...` checks with the correct label to URL conversion internally.

I'm trying to understand why `metricSet.FilterSample` needs to know there are multiple checks taking place if it only one needs to pass to allow access to the metric?

This feels like it would be a better separation of concerns to have a specialised temporary function be concerned about checking the two `userHas...` calls, so something like this:

```go
metricSet.FilterSamples(func(labels map[string]string) bool {
          if labels["project"] != "" && userHasProjectPermission(entity.ProjectURL(labels["project"]))
                return true
          }
			return userHasServerPermission(entity.ServerURL())
})
```

Reply:
I see your point. Haven't thought about it this way to be honest but it frees up lots of the logic on the filter side which shouldn't be required as this could be done by the calling end. Let me improve this, thanks.

Reply:
Fixed.